### DISCUSSION TITLE
How do you make a 'animation curve' Curve control?

### DISCUSSION THREAD

**Discussion by MechWarrior99** [AvaloniaUI/Avalonia#14117] [Q&A] [answered]
Hello, I am working on a 'animation curve' and a control to go with it. An animation curve, meaning a series of keyframes with a value and in/out tangents that determine the interpolation from one frame to the next. But I am still rather new to Avalonia and it feels like something isn't quite clicking for me.

Example of the type of curve control I want to make:
![curve](https://github.com/AvaloniaUI/Avalonia/assets/8076495/06f65310-79d7-4c82-9848-a2801865baa8)

So far I have a class that inheris from `Control`, with a `IMyCurve` property for binding the curve. I override the `Render(..)` method to draw the main curve, by adding a lot of points to a `PolyLineSegment` (couldn't find a nicer way to draw it, I'm all ears if there is).

The part I think I am having the most trouble with is how to handle the keyframes/nodes/control points. Do you use an `ItemsControl`? Do you draw and handle the input all in the main class? Do you make another control for them and add them from the main control class?

Any advice on like a high-ish level overview of how to tacke this, samples, docs, or other sources to look at would be most apprecciated!

Thanks!

**Comment by thevortexcloud** [comment]
> The part I think I am having the most trouble with is how to handle the keyframes/nodes/control points.

Adorners might work for this, but I am not sure.

**Comment by damian-666** [comment]
A custom list box wirh a canvas as a main panel.   map a grip with
dependent property Time  and Angle  or offset to the keyframe  vector2
position .  Use a mutilple valueconverter to get to the canvas space .
 Pit the  visual scale in it.  Put   and interpolate using audio function
type Splines , f(t).   The Zero tanglent can be market on some keyframes
or there target angles.   draw that path. Use that to get your target
angles.    Put the key frame times in a observableKeframecollection, sort
by time.  Put those    Each animation, Is Named key collection, must have
unique names,     has a list or map of keyframescollectuon.   Bind the
visual item template or data template for the type, to the keyprames  ,
set itemssource on the list box or list view  to the Named animation,  Draw
or interpolate ,  using the spline.  This is also in tweening code in NeZ.
  The chatgpt can code this for you .    I've done it on Wpf took a months
with a team, and the bot did a decent job in 4 hours.   I have the code and
prompts somewhere.   But not time to spend.   I think it's the Wpf way.
 If you want to Collab on if I can consult.   It's a good way to test
avalonia.


Put a filter to draw some or all the angles..   how to handle blends ,
missing data ids, , is part of general animation of models since they
change..   you don't want to loose your keyframes wehn you add  a leg .

With blending
this is a 3d studio level feature , it's needs  easy  customizability, cant
be overlay generallized.     This also related to bones and other.   Ive
spet Years on it ,I know how to prompt.

   But don't have the time to reimplembt.   I'm staying with Wpf as the
bugs are hard to iron out and I'm now sure how stable is avalonia yet ,
need docking, work for free on for sciencetists ,  I'm tempted to rewrite
it tho.  It's was  for a game level editor IDE.   I don't mind if it's open
source it's general purpose to some degree...

 Someone might have a more data separate implementation.   No state in the
vew just a view..   undoable is in mine but I'm using a old data list  way
to do that. I'm not sure it modern.  The notification  is tricky.   You
have to resort.  There is a time slider bound to the current time of the
animation.    This took months.  If You have time to experiment with the
bot Chatgpt4 I did it through the front interface special prompt or codex
.  If you set up a vault I can help .. the bot will not alwasz code for
you.   It has to like you.. take long to  earn that. It's has to approve of
your mission. Mine is nonprofit for scientist visualization research so it
will usually codee for me.

I'll need one later that is avaolia only so I don't mind helping with the
design.  It's game engine stuff but needs no ECS, or direct depends.   If
we'll designed.  It should use vector2 , time and one parameter to
animate.   And system.numerics.   that has your splines I think.. should
solve and interpolate for you.   Also useful for auto filters design.
 Since this is a fit that has only on value for each time.

**Reply by thevortexcloud** [reply]
>  I'm now sure how stable is avalonia yet ,

In my experience, it's very stable (and based on the number of complaints I see people make, much more stable than MAUI). But I can't say I have made any custom advanced controls like this. But other people have. For example: There are people who are using it as a full 3D game editor.

> need docking

There are multiple third party libraries that add docking support for Avalonia. EG https://github.com/wieslawsoltes/Dock

On a side note, I just noticed this app exists: https://github.com/stakira/OpenUtau/tree/master

It seems to have some sort of curve editor in it already so you can probably use it as the starting point for your code. It's MIT licensed so you should legally be in the clear. There are also several other apps listed here that seem to have some sort of curve editor in them: https://avaloniaui.net/Showcase

Most of them are open source.

**Comment by MechWarrior99** [comment, marked as answer]
Thanks for the suggestions and advice @thevortexcloud, @damian-666. I ended up separating it in to a number of controls combined together. I have a main `CurveView` which is just the container basically.
I then have a `CurvePath` Control which just renders the line of the animation curve.
The most important one is the `CurveKeyframePanel` which is a panel control. It has a list of keyframes that can be bound to. Importantly there is a `ItemsSourceView` private field, and when the keyframe list property changes, I set the `ItemsSourceView`
```sc
private ItemsSourceView<ICurveKeyframe> _keyframesView;

// Just anormal StyledProperty implementation....
public static readonly StyledProperty<IReadOnlyList<ICurveKeyframe>> KeyframesProperty = ...
public IReadOnlyList<ICurveKeyframe> Keyframes { get.. set..} 

protected override void OnPropertyChanged(AvaloniaPropertyChangedEventArgs change)
{
      base.OnPropertyChanged(change);

      if (change.Property == KeyframesProperty)
      {
          _keyframesView = ItemsSourceView.GetOrCreate(change.GetNewValue<IReadOnlyList<ICurveKeyframe>>());
          _keyframesView.CollectionChanged += OnKeyframesChanged;
      }
}
 ```
In `OnKeyframesChanged` I add/remove KeyframeControls as needed. And I override `ArrangeOverride` to position the keyframe controls where they should be.

And in the main `CurveView` control I just overlay the `CurvePath` and `CurveKeyframePanel` over each other.
It works for me, but overall this doesn't really feel like the 'right way`, or the `Avalonia way`. Not fully at least. So if some has more suggestions on a better approach I am all ears. But maybe this wll help someone in the future too! :)
